### ISSUE TITLE
Nightly rustc panics when the closure inside unwrap_or_else returns certain incorrect types

### ISSUE THREAD

**Issue by villuna** [rust-lang/rust#117445]
### Code

```Rust
fn main() {
    let x = Some(1).unwrap_or_else(|| Some(2));
}
```

This is the simplest version of the code I could come up with that causes the panic. From my testing I've found it does this for Option and Result, but not structs/enums I defined myself. E.g. this doesn't panic and gives the correct error message:

```Rust
enum MyOption<T> {
    Some(T),
    None,
}

fn main() {
    let x = Some(1).unwrap_or_else(|| MyOption::Some(2));
}
```

I also tried returning some other incorrect types (float and bool and &str) and none of those panic. So far I've only found this panic happens for the Option and Result types.

### Meta

`rustc +nightly --version --verbose`:
```
rustc 1.75.0-nightly (df871fbf0 2023-10-24)
binary: rustc
commit-hash: df871fbf053de3a855398964cd05fadbe91cf4fd
commit-date: 2023-10-24
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.3

```

### Error output

```
thread 'rustc' panicked at compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:1755:62:
already mutably borrowed: BorrowError

...

error: internal compiler error: no errors encountered even though `delay_span_bug` issued

error: internal compiler error: coercion error but no error emitted
  --> src/main.rs:25:39
   |
25 |     let x = Some(1).unwrap_or_else(|| Some(2));
   |                                       ^^^^^^^
   |
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0: rust_begin_unwind
             at /rustc/df871fbf053de3a855398964cd05fadbe91cf4fd/library/std/src/panicking.rs:597:5
   1: core::panicking::panic_fmt
             at /rustc/df871fbf053de3a855398964cd05fadbe91cf4fd/library/core/src/panicking.rs:72:14
   2: core::cell::panic_already_mutably_borrowed
             at /rustc/df871fbf053de3a855398964cd05fadbe91cf4fd/library/core/src/cell.rs:770:5
   3: <rustc_hir_typeck::fn_ctxt::FnCtxt>::suggest_missing_unwrap_expect
   4: <rustc_hir_typeck::fn_ctxt::FnCtxt>::emit_type_mismatch_suggestions
   5: <rustc_hir_typeck::fn_ctxt::FnCtxt>::emit_coerce_suggestions
   6: rustc_hir_typeck::check::check_fn
   7: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_closure
   8: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
   9: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_argument_types
  10: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  11: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_decl
  12: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_block_with_expected
  13: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  14: rustc_hir_typeck::check::check_fn
  15: rustc_hir_typeck::typeck
      [... omitted 1 frame ...]
  16: rustc_hir_analysis::check_crate
  17: rustc_interface::passes::analysis

...

note: delayed at compiler/rustc_hir_typeck/src/coercion.rs:1575:34
         0: <rustc_errors::HandlerInner>::emit_diagnostic
         1: <rustc_session::session::Session>::delay_span_bug::<rustc_span::span_encoding::Span, &str>
         2: rustc_hir_typeck::check::check_fn
         3: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_closure
         4: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
         5: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_argument_types
         6: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
         7: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_decl
         8: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_block_with_expected
         9: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
        10: rustc_hir_typeck::check::check_fn
        11: rustc_hir_typeck::typeck
        12: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 8]>>
        13: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 8]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, true>
        14: rustc_query_impl::query_impl::typeck::get_query_incr::__rust_end_short_backtrace
        15: rustc_hir_analysis::check_crate
        16: rustc_interface::passes::analysis
        17: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 1]>>
        18: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::SingleCache<rustc_middle::query::erase::Erased<[u8; 1]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, true>
        19: rustc_query_impl::query_impl::analysis::get_query_incr::__rust_end_short_backtrace
        20: std::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>
        21: <<std::thread::Builder>::spawn_unchecked_<rustc_interface::util::run_in_thread_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
        22: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
                   at /rustc/df871fbf053de3a855398964cd05fadbe91cf4fd/library/alloc/src/boxed.rs:2007:9
        23: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
                   at /rustc/df871fbf053de3a855398964cd05fadbe91cf4fd/library/alloc/src/boxed.rs:2007:9
        24: std::sys::unix::thread::Thread::new::thread_start
                   at /rustc/df871fbf053de3a855398964cd05fadbe91cf4fd/library/std/src/sys/unix/thread.rs:108:17
        25: <unknown>
        26: <unknown>
```

Here is the file that cargo asked me to attach: [rustc-ice-2023-10-31T12_32_40-14387.txt](https://github.com/rust-lang/rust/files/13216353/rustc-ice-2023-10-31T12_32_40-14387.txt)

</p>
</details>

**Comment by compiler-errors** [comment]
Please update your nightly compiler -- I think this is fixed

**Comment by villuna** [comment]
Oh my gosh, im so sorry i should have updated first. It is indeed fixed, thank you
